lua/SlurmTree.lua

```diff
@@ -0,0 +1,91 @@
+local ipc = require 'libipc'
+local Tree = require 'ipc.Tree'
+local NullTree = require 'ipc.NullTree'
+
+local function SlurmTree(fn, tasksPerGpu)
+   tasksPerGpu = tasksPerGpu or 1
+   local slurmProcId = tonumber(os.getenv("SLURM_PROCID"))
+   local numNodes = tonumber(os.getenv("SLURM_NTASKS"))
+   local slurmNNodes = tonumber(os.getenv("SLURM_JOB_NUM_NODES"))
+   local tasksPerHost = math.ceil(numNodes / slurmNNodes)
+   local nodeIndex = slurmProcId + 1
+
+   fn = fn or os.getenv("HOME")..'/.torch'
+   fpath = fn..'/slurm.'..os.getenv("SLURM_JOBID")..'.server'
+   local function publish(host, port)
+      os.execute('mkdir -p '..fn)
+      local f = io.open(fpath, 'w')
+      f:write(host..':'..port)
+      f:close()
+   end
+   local function query()
+      while true do
+         local f = io.open(fpath, 'r')
+         if f then
+            local s = f:read('*all')
+            if type(s) == 'string' then
+               local p = s:split(':')
+               if type(p) == 'table' and #p == 2 then
+                  return p[1], tonumber(p[2])
+               end
+            end
+            f:close()
+         end
+      end
+   end
+   local function rcsvAllPairs(b, n, i, d, f)
```

Comment:
Can you provide better names here? I just noticed that the original code has also 5 arguments with 1 letter each. We have to change that or the code becomes much harder to understand.